godot-core/src/builtin/variant/mod.rs

```diff
@@ -288,25 +301,70 @@ impl Variant {
             return &[];
         }
 
-        let variant_ptr_array: &'a [sys::GDExtensionConstVariantPtr] =
-            std::slice::from_raw_parts(variant_ptr_array, length);
-
-        // SAFETY: raw pointers and references have the same memory layout.
+        // raw pointers and references have the same memory layout.
         // See https://doc.rust-lang.org/reference/type-layout.html#pointers-and-references-layout.
-        unsafe { std::mem::transmute(variant_ptr_array) }
+        let variant_ptr_array = variant_ptr_array.cast::<&Variant>();
+
+        // SAFETY: `variant_ptr_array` isn't null so it is safe to call `from_raw_parts` on the pointer cast to `*const &Variant`.
+        unsafe { std::slice::from_raw_parts(variant_ptr_array, length) }
     }
 
-    /// Converts to variant mut pointer; can be a null pointer.
-    pub(crate) fn ptr_from_sys_mut(variant_ptr: sys::GDExtensionVariantPtr) -> *mut Variant {
-        variant_ptr as *mut Variant
+    /// Convert an array of `Variant` sys pointers to a slice of mutable `Variant` references.
+    ///
+    /// # Safety
+    ///
+    /// Either `variant_array` is null, or it must be safe to call [`std::slice::from_raw_parts`] with
+    /// `variant_array` cast to `*const Variant` and `length`.
+    pub(crate) unsafe fn borrow_var_slice<'a>(
+        variant_array: sys::GDExtensionConstVariantPtr,
+        length: usize,
+    ) -> &'a [Variant] {
+        sys::static_assert_eq_size_align!(Variant, sys::types::OpaqueVariant);
+
+        // Godot may pass null to signal "no arguments" (e.g. in custom callables).
+        if variant_array.is_null() {
+            debug_assert_eq!(
+                length, 0,
+                "Variant::unbounded_refs_from_sys(): pointer is null but length is not 0"
+            );
+            return &[];
+        }
+
+        // raw pointers and references have the same memory layout.
```

Comment:
See above

Reply:
Just realized that these actually don't need this note in this or the next one, since we're not transmuting a pointer to a reference here. We're turning a pointer into a `*const Variant` here but in the previous one we turn it into a `*const &Variant`.